### DISCUSSION TITLE
Rule/Script to repeat a command (with parameter): rfraw

### DISCUSSION THREAD

**Discussion by nagyrobi** [arendst/Tasmota#11273] [Ideas]
Occasionally some of my RF covers miss the code sent by `rfraw` command (Sonoff RF Bridge). After some intensive research and lots of reading the conclusion is that the only solution is to do the same thing as normal remote controls do, send the same code multiple times.

The most elegant way would be not to send them twice from the home automation system, as that would require to re-code the whole implementation to generate multiple events for a single event etc..., but instead make a rule or a script in Tasmota which would trigger when an `rfraw` comes through MQTT and the value is not `0`, execute `rfraw` again with the same value.

Currently the system sends the command to Tasmota via MQTT by sending to the topic `cmnd/rf-bridge-1/backlog` the payload `'rfraw AAB035050412FC061802C6018622B0A481A3B2B2A3B2A3A3B2B2B2A3A3B2B2A3B2A3B2A3A3B2A3A3B2A3B2A3A3B2B2A3B2B2A3B2A3B2A3B255;rfraw 0'`.

So the goal would be to catch this before execution, execute the first `rfraw` twice and afterwards the final `rfraw 0`.
I don't mind changing the topic if required, moreover, I think I could live with any different payload (for example containing only the code) and add the rfraws and the 0 by the script/rule.

Any hints? Maybe with `event`?

**Comment by meingraham** [comment]
Have your home automation system send an `Event` with the data rather than the `RFRaw` command itself. Then, when the event is triggered it performs the multiple `RFRaw` with the data sent by the HAB followed by the `RFRaw 0`.

**Reply by nagyrobi** [reply]
I can have the automation system send an `event` with the data. I need some help to create the rule as you suggest.

**Reply by barbudor** [reply]
If you have your H.A. send to `cmnd/rf-bridge-1/event` payload 'rfrf=AAB03....3B2A3B255'
The rule would be
`on event#rfrf do backlog RfRaw %value%; RfRaw %value%; RfRaw 0 endon`

**Reply by nagyrobi** [reply]
Unfortunately it doesn't work. It seems that the payload gets truncated to `AAB03505141306060E02C601902378A481A3B2B2A3B2A3A3B2B2B2A3A3B2B2A3B2A3B2A3A3B2A3A3B2A3B2A3A3B2B2...` The RF brigde just beeps.

**Comment by nagyrobi** [comment]
Or should I instead [modify the code](https://github.com/arendst/Tasmota/issues/5936#issuecomment-500236581) in the payloads to make the Portisch firmware do the repetitions? That will really repeat?

**Reply by meingraham** [reply]
Not sure every situation warrants always transmitting the code multiple times. And if it does, is two times the right number? Three times? Tt's relatively simple to do it via a rule. So, I would say no.

**Reply by nagyrobi** [reply]
Well it looks like the codes prepared by [BitBucketConverter.py](https://github.com/Portisch/RF-Bridge-EFM8BB1) are already repeated 4 times by default (I wasn't aware of this). So changing the repetition to, say, 20, as stated in the example above could be one approach. The transmission repeats are done by the onboard EFM8BB1 chip in the RF Bridge based on the code in the payload - which means less stress on the Tasmota part of things.

So if I understand correctly, sending the code as it is now, means it's already repeated 4 times, and that doesn't seem to be enough. I could send it with 2 rfraw commands and that would be equal to 8 repetitions in 2 separate transmissions, or by changing a byte, that would be 20 repetitions in 1 transmission. I'll try this first.

**Reply by nagyrobi** [reply]
Ok, tested, reliability got even worse after changing that byte. Back to the original idea.

**Reply by nagyrobi** [reply]
Created https://github.com/arendst/Tasmota/issues/11296 issue as there's a bug somewhere...

**Comment by nagyrobi** [comment]
@gemu2015 can you please give me a hand on this?
The [idea](https://github.com/arendst/Tasmota/discussions/11273#discussioncomment-456320) was to go with rules but that doesn't work because variables in rules truncate the string size and corrupts the data. I need around 125-130 chars in a variable.

**Reply by sfromis** [reply]
As described in the other [thread](https://github.com/arendst/Tasmota/issues/11296), it is simple to circumvent that limitation.

**Reply by gemu2015** [reply]
in scripts the default max string size is set to
#define SCRIPT_MAXSSIZE 48
you may change this value to up to 255 in scripter source code  (keep it as low as needed)

i will make this configurable in config override in a future release:

#ifndef SCRIPT_MAXSSIZE
#define SCRIPT_MAXSSIZE 48
#endif

this will however also have an impact on heap memory and stack

**Reply by nagyrobi** [reply]
Any tip on the script itself, how to assign a variable from an MQTT topic?

**Reply by gemu2015** [reply]
you may try this

\>D 150
lstr=""
\>E
lstr=rfraw

**Reply by nagyrobi** [reply]
What's the MQTT topic watched by the script?

Note it's not the RESULT we need, we want to catch first what's in the payload of the MQTT message and run commands with it.

So for example in a topic `cmnd/rf-bridge-2/something` payload `AAB035050412FC060E02C6018622CEA481A3B2B2A3B2A3A3B2B2B2A3A3B2B2A3B2A3B2A3A3B2A3A3B2A3B2A3A3B2B2A3B2B2B2B2A3B2B2B255` appears, the script should execute:
```
rfraw AAB035050412FC060E02C6018622CEA481A3B2B2A3B2A3A3B2B2B2A3A3B2B2A3B2A3B2A3A3B2A3A3B2A3B2A3A3B2B2A3B2B2B2B2A3B2B2B255
rfraw AAB035050412FC060E02C6018622CEA481A3B2B2A3B2A3A3B2B2B2A3A3B2B2A3B2A3B2A3A3B2A3A3B2A3B2A3A3B2B2A3B2B2B2B2A3B2B2B255
rfraw 0
```
Rule can't do it because the payload is too big to fit in a variable (only by about 30 chars). @sfromis suggested a dirty hack to split it etc but it's harder to maintain on the HA side so if rules can't be patched to fit these in, I'd rather go with scripting.

**Reply by nagyrobi** [reply]
@gemu2015 I definitely would like to test the script engine with the same task because we finally made rules work but delay is really noticeable. Covers start moving about half a second later than before so it looks like those repeated backlogs and var sets require processing time whics adds up and is not negligible. I hope scripts are more fast.

**Reply by sfromis** [reply]
You may be able to improve the speed by trimming down the backlog inter-command delay [SetOption34](https://tasmota.github.io/docs/Commands/#setoption34) from the default of 200 milliseconds.

**Comment by nagyrobi** [comment]
For readers coming here via a search, see https://github.com/arendst/Tasmota/issues/11296#issuecomment-798926139, repeating the code is useless.
